core/src/main/java/com/yahoo/oak/NativeMemoryManager.java

```diff
@@ -106,21 +108,52 @@ public boolean isReferenceDeleted(long reference) {
         return rcmm.isReferenceDeleted(reference);
     }
 
+    @Override public boolean isReferenceValidAndNotDeleted(long reference) {
+        return rcmm.isReferenceValidAndNotDeleted(reference);
+    }
+
     @Override
     public boolean isReferenceConsistent(long reference) {
         return rcmm.isReferenceConsistent(reference);
     }
 
+    @Override
+    public Slice getEmptySlice() {
+        return new Slice(OFF_HEAP_HEADER_SIZE, rcmm.getInvalidReference());
+    }
+
+    @VisibleForTesting
+    @Override
+    public int getHeaderSize() {
+        return OFF_HEAP_HEADER_SIZE;
+    }
+
     @Override
     public long allocated() {
         return allocator.allocated();
     }
 
+    // Native memory manager requires metadata header to be placed before the user data written off-heap
+    // therefore the bigger than requested size is allocated
     @Override
-    public void allocate(Slice s, int size) {
-        boolean allocated = allocator.allocate(s, size);
+    public void allocate(Slice s, int size, boolean existing) {
```

Comment:
I have some reservations that the flag that hints about replacing existing value, implicitly sets the lock state.
This should be at least be addressed in the documentation.
If you have in mind a design that avoids that flag it is even better, but I guess this can wait for future PRs.

Reply:
The "existing" boolean flag parameter is explained in interface method in Memory Manager interface file. However, indeed doesn't explain that this will cause the lock to be treated differently. This is because it is internal implementation and after all (after allocation) the lock is released, this way or the other... I didn't plan to implement it differently, but you are welcome to suggest when we will discuss the  Memory Manager interface for putting it public.

Reply:
If you want we can discuss it further

Reply:
I don't think this requires a discussion. I only made a suggestion to add a comment to this implementation explaining the lock thing.

Reply:
I have added the comment. Thanks